src/Lean/Server/Rpc/Deriving.lean

```diff
@@ -134,115 +110,93 @@ private def deriveStructureInstance (indVal : InductiveVal) (params : Array Expr
-          let tm := Syntax.mkApp (mkIdent <| Name.mkStr tgt ctor.getString!) items
-          `(return $tm:term)
-
-        let encArm ← `(matchF| | $(mkPattern indVal.name):term => $(← mkBody packetNm ``rpcEncode))
-        let decArm ← `(matchF| | $(mkPattern packetNm):term => $(← mkBody indVal.name ``rpcDecode))
-
-        return { acc with ctors := acc.ctors.push pktCtor
-                          encodes := acc.encodes.push ⟨encArm⟩
-                          decodes := acc.decodes.push ⟨decArm⟩ }
-
-      -- helpers for type name syntax
-      let paramIds ← params.mapM fun p => return mkIdent (← getFVarLocalDecl p).userName
-      let typeId := Syntax.mkApp (← `(@$(mkIdent indVal.name))) paramIds
-      let packetAppliedId := Syntax.mkApp (mkIdent packetNm) (st.uniqEncArgTypes.map (mkIdent ·))
-
-      `(variable $st.binders*
-
-        inductive $(mkIdent packetNm) where
-          $[$(st.ctors):ctor]*
-          deriving $(mkIdent ``FromJson), $(mkIdent ``ToJson)
-
-        instance : $(mkIdent ``RpcEncoding) $typeId $packetAppliedId where
-          rpcEncode := fun x => match x with
-            $[$(st.encodes):matchAlt]*
-          rpcDecode := fun x => match x with
-            $[$(st.decodes):matchAlt]*
-      )
+      let encTp ← getRpcPacketFor argTp
+      pktCtorTp := mkForall (← getFVarLocalDecl arg).userName BinderInfo.default encTp pktCtorTp
+    -- TODO(WN): this relies on delab printing non-macro-scoped user names in non-dependent foralls
+    -- to generate the expected JSON encoding
+    let pktCtorTpStx ← PrettyPrinter.delab pktCtorTp
+    let pktCtor ← `(Lean.Parser.Command.ctor| | $(mkIdent ctor.getString!):ident : $pktCtorTpStx:term)
+
+    -- create encoder and decoder match arms
+    let nms ← argVars.mapM fun _ => mkIdent <$> mkFreshBinderName
+    let mkPattern (src : Name) := Syntax.mkApp (mkIdent <| Name.mkStr src ctor.getString!) nms
+    let mkBody (tgt : Name) (func : Name) : TermElabM Term := do
+      let items ← nms.mapM fun nm => `(← $(mkIdent func) $nm)
+      let tm := Syntax.mkApp (mkIdent <| Name.mkStr tgt ctor.getString!) items
+      `(return $tm:term)
+
+    let encArm ← `(matchF| | $(mkPattern indVal.name):term => $(← mkBody packetNm ``rpcEncode))
+    let decArm ← `(matchF| | $(mkPattern packetNm):term => $(← mkBody indVal.name ``rpcDecode))
+
+    return { acc with ctors := acc.ctors.push pktCtor
+                      encodes := acc.encodes.push ⟨encArm⟩
+                      decodes := acc.decodes.push ⟨decArm⟩ }
+
+  -- helpers for type name syntax
+  let paramIds ← params.mapM fun p => return mkIdent (← getFVarLocalDecl p).userName
+  let typeId := Syntax.mkApp (← `(@$(mkIdent indVal.name))) paramIds
+  let packetId := Syntax.mkApp (mkIdent packetNm) packetParamNames
+
+  `(inductive $(mkIdent packetNm) $packetParamBinders:bracketedBinder* where
+      $[$(st.ctors):ctor]*
+      deriving FromJson, ToJson
+
+    variable $(paramBinders ++ packetParamBinders ++ encInstBinders)* in
```

Comment:
The point of using `variable` was to ensure that we can have `RpcEncoding` without requiring a packet for ghost parameters. For example, we should have `RpcEncoding (TSyntax n) _` without `RpcEncoding n _` because the `variable` will not be introduced if it doesn't appear in the body. Could you please check that this works and add a ghost case to the test? I clearly overdid it with the `variable`s though, we don't need them for *all* the types appearing inside.

Technically, if I understand the code correctly without having read it in detail (`derive` handlers are gnarly :/) the semantics after this PR will be different - we need `RpcEncoding` for argument types *at definition time*, whereas before we did not. In practice this leads to delayed synthesis failures, so the eager approach is better.

Reply:
> For example, when `inductive Unused (α : Type)`, we should have `RpcEncoding (Unused α) _` without `RpcEncoding α _` because the `variable [RpcEncoding α _]` will not be introduced if `@rpcEncode α` doesn't appear in the body. Could you please check that this works and add a ghost case to the test?

(I've changed the name because we already have a different type called `PEmpty`.)

This doesn't work, but for a different reason.  Namely the type of the instance is `(α αPacket : Type) →
  RpcEncoding (Unused α) (Unused.RpcEncodingPacket αPacket)`.  The instance doesn't have a `RpcEncoding α _` argument, but it will never apply because TC synthesis can't infer the `αPacket` argument.

Another issue is that empty inductives don't work.

Do you have a concrete example where you need unused parameters?

> the semantics after this PR will be different - we need `RpcEncoding` for argument types at definition time, whereas before we did not.

Right.

> In practice this leads to delayed synthesis failures, so the eager approach is better.

Yes, I was very surprised when I tried do derive `RpcEncoding` for an inductive type and then type class synthesis couldn't find it.

Reply:
Both issues are fixed now.

Reply:
> Do you have a concrete example where you need unused parameters?

I was thinking of `TSyntax`, but there the index is not a type and we already skip non-Type indices anyway. Actually this technically makes the instance unsound, because `TSyntax n` and `TSyntax m` have the same encoding and we can do ``rpcDecode (α := TSyntax `abc) (rpcEncode (α := TSyntax `def) stx)``. Not sure whether this is actually an issue.

Reply:
Note that the `TSyntax` instance is only unsound in the sense that it can violate the (not particularly enforced) invariant that a `TSyntax ks` has one of the "kinds" `ks`.  But I don't think there's any realistic way to enforce that invariant, even if we serialized `ks` as well.

It is not memory unsafe though.